beater/pub.go

```diff
@@ -52,44 +61,47 @@ func newPublisher(pipeline beat.Pipeline, N int, shutdownTimeout time.Duration)
 
 		// Set channel size to N - 1. One request will be actively processed by the
 		// worker, while the other concurrent requests will be buffered in the queue.
-		events: make(chan []beat.Event, N-1),
+		decodedRequests: make(chan decodedRequest, N-1),
+	}
+
+	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
+		go p.run()
 	}
 
-	go p.run()
 	return p, nil
 }
 
 // Stop closes all channels and waits for the the worker to stop.
-// The worker will drain the queue on shutdown, but no more events
+// The worker will drain the queue on shutdown, but no more decodedRequests
 // will be published.
 func (p *publisher) Stop() {
 	p.m.Lock()
 	p.stopped = true
 	p.m.Unlock()
-	close(p.events)
+	close(p.decodedRequests)
 	p.client.Close()
 }
 
-// Send tries to forward events to the publishers worker. If the queue is full,
+// Send tries to forward decodedRequests to the publishers worker. If the queue is full,
 // an error is returned.
 // Calling send after Stop will return an error.
-func (p *publisher) Send(batch []beat.Event) error {
+func (p *publisher) Send(decodedRequest decodedRequest) error {
 	p.m.RLock()
 	defer p.m.RUnlock()
 	if p.stopped {
-		return errChanneClosed
+		return errChannelClosed
 	}
 
 	select {
-	case p.events <- batch:
+	case p.decodedRequests <- decodedRequest:
 		return nil
 	case <-time.After(time.Second * 1): // this forces the go scheduler to try something else for a while
 		return errFull
 	}
 }
 
 func (p *publisher) run() {
-	for batch := range p.events {
-		p.client.PublishAll(batch)
+	for decodedRequest := range p.decodedRequests {
+		p.client.PublishAll(decodedRequest.payload.Transform(decodedRequest.config))
```

Comment:
this works but reads strangely (to me).  What do you think of flipping it around to `func (dr *decodedRequest) Publish(client beat.Client)` ?  I'm not sure that's any better.

At least, can we rename the `decodedRequest` iterator to something distinct from the `decodedRequest` struct type?

Reply:
The `client.PublishAll([] beat.Event)` is defined in beats, so flipping it here would only defer calling this method, which I think is even more confusing. 
I tried to come up with a better naming though. WDYT?

Reply:
Yeah, and after sleeping on it, it's even more gross to have the decodedRequest know anything about publishing anyway.  I like rename.